### ISSUE TITLE
Extending HB-E Boundary Conditions?

### ISSUE THREAD

**Issue by ed-p-may** [ladybug-tools/honeybee-energy#1049]
Hi, @chriswmackey , @mostaphaRoudsari ,

I have a question related to extending HB-Energy [Boundary Condition](https://github.com/ladybug-tools/honeybee-energy/blob/master/honeybee_energy/boundarycondition.py) objects:

I am trying to look at what it would take to implement a ['Kiva'](https://kiva.readthedocs.io/en/stable/) style Foundation BC within our Honeybee-REVIVE extension. In order to add this new BC, the first step is to create a new custom BC. This part I understand 
based on how you did it for "Adiabatic" and "OtherSideTemperature" already, and it appears that I am able to easily extend hb-energy with a new "Foundation" BC object. I can set HB-face BCs to this new type, and it writes out to IDF as expected (`Foundation, !- Outside Boundary Condition`). So that part is great. 

However: in addition to setting the BC name to "Foundation", I also need to set the `Outside Boundary Condition Object` to be a reference to a new Kivy object (which will be appended to the IDF elsewhere....).

Reviewing the honeybee-energy code, I *think* (maybe) I understand where the BC IDF text is being generated, and based on my reading here it appears that there is no way to add this new reference? It seems that only the special types "Surface" and "OtherSideTemperature" add the bc-object text to the idf string?

from [honeybee_energy.writer.face_to_idf](https://github.com/ladybug-tools/honeybee-energy/blob/df6c999f2a5c1f82e7f373f1658cc1b6b7fe8a61/honeybee_energy/writer.py#L434)
```python
. . . 
# select the correct boundary condition
bc_name, append_txt = face.boundary_condition.name, None
if isinstance(face.boundary_condition, Surface):
    face_bc_obj = face.boundary_condition.boundary_condition_object
elif face.boundary_condition.name == 'OtherSideTemperature':
    face_bc_obj = '{}_OtherTemp'.format(face.identifier)
    append_txt = face.boundary_condition.to_idf(face_bc_obj)
    bc_name = 'OtherSideCoefficients'
else:
    face_bc_obj = ''
. . . 
```

So given the above, do you think it is accurate to say it is not possible to have a new custom BC add text to the "Outside Boundary Condition Object" line? Perhaps I am missing something though?

I suppose that I could subclass 'Surface' with my new BC and that would get past the `isinstance` check?

But I wonder if the above could be made a bit more extensible and allow any BC that has a `.boundary_condition` attribute to write itself to the IDF?

Perhaps something like below, instead of checking the nominal type of `Surface`, just check for the bc attribute?

```python
. . . 
# select the correct boundary condition
bc_name, append_txt = face.boundary_condition.name, None
if face.boundary_condition.name == 'OtherSideTemperature':
    face_bc_obj = '{}_OtherTemp'.format(face.identifier)
    append_txt = face.boundary_condition.to_idf(face_bc_obj)
    bc_name = 'OtherSideCoefficients'
elif hasattr(face.boundary_condition, "boundary_condition_object"):  #<----
   face_bc_obj = face.boundary_condition.boundary_condition_object   #<-----
else:
    face_bc_obj = ''
. . . 
```

any thoughts are much appreciated as always,
@ed-p-may

**Comment by chriswmackey** [comment]
Hey @ed-p-may ,

You should be thinking of KIVA boundary conditions this as something that goes completely in honeybee-energy. Because it's native E+ functionality, it should live in this package and not an extension.

With that in mind, what stops you from treating this like a Surface boundary condition? Do you need to references lots of different types of "Outside Boundary Condition Objects?" Or is it just one that gets used everywhere?

**Comment by ed-p-may** [comment]
Hi @chriswmackey , interesting. I'd be all for adding it as a first-class HBE object. Happy to help put that together.

From what I understand (still learning how KIVA works....) a very basic working implementation would include 3 elements in the IDF:


## 1) A new "Foundation" object:
```
FOUNDATION:KIVA,
  My Kiva Foundation,      !- Name
  ,                        !- Initial Indoor Air Temperature {C}
  ,                        !- Interior Horizontal Insulation Material Name
  0,                       !- Interior Horizontal Insulation Depth {m}
  ,                        !- Interior Horizontal Insulation Width {m}
  ,                        !- Interior Vertical Insulation Material Name
  ,                        !- Interior Vertical Insulation Depth {m}
  ,                        !- Exterior Horizontal Insulation Material Name
  ,                        !- Exterior Horizontal Insulation Depth {m}
  0,                       !- Exterior Horizontal Insulation Width {m}
  EPS 2in,                 !- Exterior Vertical Insulation Material Name #<--- THIS MATERIAL ALSO NEEDS TO BE ADDED
  0.0003048,               !- Exterior Vertical Insulation Depth {m}
  0.2,                     !- Wall Height Above Grade {m}
  0,                       !- Wall Depth Below Slab {m}
  ,                        !- Footing Wall Construction Name
  ,                        !- Footing Material Name
  0.4;                     !- Footing Depth {m}
```

## 2) A new "Properties" object:
```
SURFACEPROPERTY:EXPOSEDFOUNDATIONPERIMETER,
  Basement,                !- Surface Name
  TotalExposedPerimeter,   !- Exposed Perimeter Calculation Method
  12.192,                  !- Total Exposed Perimeter {m}
  1,                       !- Exposed Perimeter Fraction {dimensionless}
  Yes;                     !- Surface Segment 1 Exposed
```

## 3) Setting the Building Surface BC values:
This has two parts: 
- Set the "Boundary Condition" to "Foundation"
- Set the "Boundary Condition Object" to reference the KIVA-Foundation object

So that looks roughly like:
```
BuildingSurface:Detailed,
  Slab,                    !- Name
  Floor,                   !- Surface Type
  Exterior Slab UnIns,     !- Construction Name
  WholeHouse,              !- Zone Name
  ,                        !- Space Name
  Foundation,              !- Outside Boundary Condition #<------- HERE. 'Surface' > 'Foundation'
  My Kiva Foundation,      !- Outside Boundary Condition Object #<------ HERE
  NoSun,                   !- Sun Exposure
  NoWind,                  !- Wind Exposure
  0,                       !- View Factor to Ground
  4,                       !- Number of Vertices
  8.1534,11.6713,0,  !- X,Y,Z ==> Vertex 1 {m}
  8.153399999999,8.013700000009,0,  !- X,Y,Z ==> Vertex 2 {m}
  0,8.013700000009,0,  !- X,Y,Z ==> Vertex 3 {m}
  0,11.6713,0;  !- X,Y,Z ==> Vertex 4 {m}
```

- - -
So, given that, the existing "Surface" BC can almost work, except that it has to yield a [`boundarycondition.name`](https://github.com/ladybug-tools/honeybee-energy/blob/df6c999f2a5c1f82e7f373f1658cc1b6b7fe8a61/honeybee_energy/writer.py#L433C5-L433C61) of "Foundation" when writing to the IDF, rather than "Surface"? So I think HBE would need a whole new type for that, since the name is set to [always be the type?](https://github.com/ladybug-tools/honeybee-core/blob/d247c34c79e9585f6ab85fef7d9f21e0d725c1c2/honeybee/boundarycondition.py#L17)

**Comment by chriswmackey** [comment]
Thanks for clarifying, @ed-p-may .

I am tempted to wrap all of the attributes of the foundation and the properties object into a singe new `GroundKIVA` boundary condition object that gets assigned to honeybee Faces. This would basically mean that each Floor face with a `GroundKIVA` boundary condition would only ever have one Foundation object and one `SURFACEPROPERTY:EXPOSEDFOUNDATIONPERIMETER` object. I'm guessing there are other types of surface property objects here, no? Are basement floors modeled the same way as a slab on grade?

In any event, I just know that everything is going to be much easier to implement on the honeybee side if everything is wrapped into a single object that can be assigned to a Face. Do you think users will find it acceptable that they may get a lot of Foundation objects in their IDF because each Floor face will get a separate foundation object?

**Comment by ed-p-may** [comment]
@chriswmackey Personally, I think it is totally fine having unique `FOUNDATION:KIVA` instances for each floor surface in a Honeybee-Room, and to merge the `FOUNDATION:KIVA` and the `SURFACEPROPERTY:EXPOSEDFOUNDATIONPERIMETER` together into a single new Honeybee-Energy object.

One thing to note is that (from what I understand, just reading [their docs](https://bigladdersoftware.com/epx/docs/24-2/engineering-reference/ground-heat-transfer-calculations-using-kiva.html#boundary-conditions) and experimenting a bit) it appears that the floor surfaces *and* the wall surfaces need to reference the same `FOUNDATION:KIVA` object. So that new Honeybee Object would need to be reference-able from multiple surfaces, I suppose similar to how constructions are handled.

From what I gather, I think that the basic KIVA IDF structure is something like this:

![Screenshot 2024-11-19 at 10 16 21 AM](https://github.com/user-attachments/assets/7a6816ae-8f6a-4148-9054-4542dd8bfec5)

Where:
1) There is a new `FOUNDATION:KIVA` object
2) The Floor(s) *and* the Walls reference this object in their `Outside Boundary Condition Object`
3) There is a new `SURFACEPROPERTY:EXPOSEDFOUNDATIONPERIMETER`
4) This references *just* the Floor surface

Do you think that changes how we would anticipate approaching these items?

**Comment by ed-p-may** [comment]
For sure! There is no time pressure on our end. We have a workaround that we can apply internally for now and no specific time-table on when we are trying to make theHoneybee-REVIVE package available to users. From [our testing so far](https://discourse.ladybug.tools/t/finding-the-difference-between-two-models/35110/4?), it looks like this mainly matters for un- or poorly-insulated ground contact surfaces (which, of course, are a big part of 'retrofit' modeling though...).

Although: I suppose a minimal implementation would be to apply the KIVA Foundation **only** to floor surfaces and just omit walls? In that case, each BC could still be unique to the single floor surface? From a quick test, this appears to show *almost* the same result as if it is applied to the walls and the floor? I suppose it depends on the extent of the subgrade walls... but I would imagine that for *most* cases this would be good enough?

![Screenshot 2024-11-19 at 4 56 51 PM](https://github.com/user-attachments/assets/9473e561-f58e-4889-b3e6-24de30432b6a)

(Note: the above is showing a 'power-outage' scenario where the mech switches 'off' for a week in winter)

model attached:
[testing_kiva.gh.zip](https://github.com/user-attachments/files/17821967/testing_kiva.gh.zip)

**Comment by chriswmackey** [comment]
Thanks, @ed-p-may .

Even if the results are close, I would prefer to do this right the first time rather than assemble something quick that I then have to change after others start using it and building on top of it. I think I can get the use of references to foundation objects to work. Now that I think about it more, maybe the Foundation should be a special type of construction object that, when it's assigned to Honeybee Faces with a Ground boundary condition, triggers the ground to be modeled with KIVA instead of the usual method.

Let me think about it more and get back to this.

**Comment by ed-p-may** [comment]
# Design Proposal: KIVA Foundation Support (#1049)

Hi @chriswmackey,

I had occasion to look back at this recently, and was working through a few design options with an LLM. I've collected together a new, revised proposal here for what this could look like. If this seems sensible to you, I'd like to take a run at implementing this:

- - - 
This proposal builds on your suggestion that the foundation should be *"a special type of object that, when it's assigned to Honeybee Faces with a Ground boundary condition, triggers the ground to be modeled with KIVA instead of the usual method"* — with one adjustment forced by E+'s referencing rules (explained below).

**TL;DR of the proposed design:**

- **No new boundary condition.** Faces keep their normal `Ground` BC.
- A new shared resource object, **`FoundationKiva`**, lives at the model level (exactly like `SHWSystem`), and is referenced by faces through a new optional **`face.properties.energy.kiva_foundation`** attribute.
- Any `Ground`-BC Floor or Wall face with a `kiva_foundation` assigned gets translated to `Outside Boundary Condition = Foundation` in the IDF/OSM. Faces without one behave exactly as they do today.
- A `FoundationKivaSettings` singleton on `ModelEnergyProperties` (like `ventilation_simulation_control`) holds the model-wide soil properties.
- Exposed perimeter is auto-calculated from geometry at translation time (with a per-face numeric override).

---

## 1. Background: what EnergyPlus actually requires

I dug through the v25.1 IDD, the Engineering Reference, and the source-level error checks in `HeatBalanceKivaManager.cc` / `SurfaceGeometry.cc`. The relevant IDF structure:

```
                     ┌───────────────────────────────┐
                     │ FOUNDATION:KIVA               │
                     │  "Basement_Foundation"        │◄──────────────────┐
                     └───────────────▲───────────────┘                   │
                                     │ Outside BC Object                 │ Outside BC Object
                     ┌───────────────┴───────────────┐   ┌───────────────┴───────────────┐
                     │ BuildingSurface:Detailed      │   │ BuildingSurface:Detailed      │
                     │  "Basement_Slab"  (Floor)     │   │  "Basement_Wall_1..n" (Wall)  │
                     │  Outside BC = Foundation      │   │  Outside BC = Foundation      │
                     └───────────────▲───────────────┘   └───────────────────────────────┘
                                     │ Surface Name
                     ┌───────────────┴───────────────┐   ┌───────────────────────────────┐
                     │ SURFACEPROPERTY:              │   │ FOUNDATION:KIVA:SETTINGS      │
                     │  EXPOSEDFOUNDATIONPERIMETER   │   │  (optional, one per model,    │
                     │  (REQUIRED, floors only)      │   │   all fields defaulted)       │
                     └───────────────────────────────┘   └───────────────────────────────┘
```

The referencing rules (all source-verified Severe errors, not just documentation conventions):

1. Only **Floor and Wall** surfaces may use the `Foundation` BC.
2. **Per zone, each foundation floor must reference a *different* `Foundation:Kiva` object** ("one floor per Foundation:Kiva per zone"). The same object *may* be reused by floors in different zones — E+ instantiates a separate 2-D Kiva model per zone.
3. Any number of walls may reference a `Foundation:Kiva`, but that object must also be referenced by **exactly one floor in the same zone**. A floor with no walls is fine (slab-on-grade).
4. Every foundation floor **requires** a `SurfaceProperty:ExposedFoundationPerimeter`. Zero exposed perimeter is legal (core-zone slab → 1-D calc against deep ground).
5. Every material referenced by `Foundation:Kiva` fields, and **every layer of the floor/wall constructions** on foundation faces, must be a regular `Material` (no `Material:NoMass` / R-only layers).
6. No windows/doors on foundation surfaces; no `Construction:InternalSource` (no radiant slabs on Kiva faces); a **weather file is required** (Kiva's accelerated initialization reads the EPW directly, so design-day-only runs fail).
7. Walls of different heights or constructions sharing one `Foundation:Kiva` are handled *inside* E+ (it splits them into separate 2-D instances and area-weights the results back) — walkout basements work without any special handling on our side.
8. A blank "Outside Boundary Condition Object" is legal (E+ auto-generates a default foundation), and `Foundation:Kiva` itself has only one required field (`Name`) — so a "just use Kiva defaults" workflow is cheap to support.

Rule 3 is the one that killed the ideas we discussed in the thread — walls and floor must share *one object* even though they have *different constructions*.

---

## 2. Why this design (and not the alternatives)

| | **A. New `Foundation` BC class** (carrying the Kiva spec, consolidated at translation) | **B. Special construction subtype** (triggers Kiva on Ground faces) | **C. Face-level reference to a shared model resource** ← proposed |
|---|---|---|---|
| Wall + floor share one `Foundation:Kiva` | Awkward — BCs have no reference/abridged mechanism, so the full spec duplicates into every face dict | **Broken** — wall and floor need *different* constructions but the *same* Kiva object | Natural — same identifier reference from any number of faces (the construction/SHW pattern) |
| honeybee-schema impact | New member in the `Face.boundary_condition` union → strict-validating consumers reject the model | Additive | Additive only: optional key on `FaceEnergyPropertiesAbridged` + optional `foundations` list on `ModelEnergyProperties` |
| Backward compatibility | `Face.from_dict` **silently degrades unknown BCs to `Outdoors`** — old cores turn basements into exterior floors | OK | Old tools see valid `Ground` faces; the model still simulates (with the standard ground method) — graceful degradation |
| Ripple into BC-aware code | `EXTERIOR_BCS` in loadbalance, TRACE preprocessing, gbXML, visualization, dragonfly, ConstructionSet BC mapping … | Construction setter type-checks, ConstructionSet logic | None — `Ground` BC everywhere; ConstructionSets already resolve these faces to `ground_construction` |
| Where Kiva params live | On the BC (conceptually fine) | On a construction (conceptually wrong — edge insulation & footings aren't construction properties) | On a dedicated resource object mirroring `Foundation:Kiva` 1:1 |

Design C is really your option B with the trigger object moved from the construction slot to its own face property, so that a slab and its basement walls (different constructions) can point at the same foundation.

The honest weaknesses of C, so they're on the table:

- **Discoverability**: "Visualize by BC" style tools will show these faces as `Ground`, not `Foundation`. I think that's acceptable (it *is* a ground face; Kiva is just the calculation method), and a validation check plus component defaults cover the main foot-gun.
- **Silent no-op risk**: a user sets `kiva_foundation` on an `Outdoors` face and nothing happens. Handled with an explicit validate-model check (see §6).

---

## 3. Object model

```mermaid
classDiagram
    class Model
    class FoundationKivaSettings {
        soil_conductivity = 1.73
        soil_density = 1842
        soil_specific_heat = 419
        ground_solar_absorptivity = 0.9
        ground_thermal_absorptivity = 0.9
        ground_surface_roughness = 0.03
        far_field_width = 40
        deep_ground_boundary = Autoselect
        deep_ground_depth = autocalculate
        minimum_cell_dimension = 0.02
        maximum_cell_growth_coeff = 1.5
        simulation_timestep = Hourly
    }
    class FoundationKiva {
        identifier
        initial_indoor_air_temperature = None
        interior_horizontal_insulation + depth/width
        interior_vertical_insulation + depth
        exterior_horizontal_insulation + depth/width
        exterior_vertical_insulation + depth
        wall_height_above_grade = 0.2
        wall_depth_below_slab = 0.0
        footing_wall_construction = None
        footing_material = None
        footing_depth = 0.3
    }
    class FaceEnergyProperties {
        construction
        vent_crack
        kiva_foundation = None
        exposed_perimeter = autocalculate
    }
    Model "1" o-- "0..*" FoundationKiva : properties.energy.foundations (collected from faces, like shws)
    Model "1" o-- "0..1" FoundationKivaSettings : properties.energy.foundation_kiva_settings
    FaceEnergyProperties "0..*" --> "0..1" FoundationKiva : kiva_foundation (by reference)
    FoundationKiva --> "0..4" EnergyMaterial : insulation and footing materials (collected into model materials)
    FoundationKiva --> "0..1" OpaqueConstruction : footing_wall_construction
```

Notes:

- **`FoundationKiva`** (new module `honeybee_energy/foundation.py`, modeled on `shw.py`): mirrors the E+ object 1:1, all fields optional/defaulted per the IDD, with `lock()`/`unlock()`, `duplicate()`, `to_dict(abridged)`/`from_dict`/`from_dict_abridged(materials, constructions)`, and `to_idf()`. Insulation/footing references are held as actual `EnergyMaterial`/`OpaqueConstruction` objects so they serialize abridged by identifier and get swept into the model's `materials`/`constructions` collections (E+ hard-errors if they're missing from the IDF).
- **All `FoundationKiva` dimensions are SI (meters) regardless of model units** — consistent with how material thicknesses already work. Only the auto-computed exposed perimeter depends on geometry, and it's evaluated after `convert_to_units()` inside the translators.
- **`FaceEnergyProperties`** gains two slots: `kiva_foundation` (default `None`) and `exposed_perimeter` (default `autocalculate`; accepts a number ≥ 0; only meaningful on Floors). Threaded through `to_dict`/`from_dict`/`apply_properties_from_dict`/`duplicate`/`is_equivalent` exactly like `construction`.
- **`ModelEnergyProperties`** gains: a `foundations` collection property (walks faces, dedupes — the `shws` pattern), and a `foundation_kiva_settings` singleton (default `None` → nothing written, E+ defaults apply — the `ventilation_simulation_control` pattern).
- **honeybee-core requires zero changes.**

### HBJSON serialization (abridged)

```jsonc
// on the Face — additive, optional keys only
{
  "type": "Face",
  "identifier": "Basement_Slab",
  "boundary_condition": {"type": "Ground"},
  "properties": {
    "energy": {
      "type": "FaceEnergyPropertiesAbridged",
      "construction": "Basement Slab Construction",
      "kiva_foundation": "Basement_Foundation",           // NEW (optional)
      "exposed_perimeter": {"type": "Autocalculate"}      // NEW (optional)
    }
  }
}

// at the model level — new optional resource list + singleton
"energy": {
  "foundations": [
    {
      "type": "FoundationKivaAbridged",
      "identifier": "Basement_Foundation",
      "exterior_vertical_insulation": "EPS 2in",          // material identifier
      "exterior_vertical_insulation_depth": 0.6,
      "wall_height_above_grade": 0.2,
      "footing_depth": 0.3
    }
  ],
  "foundation_kiva_settings": { "type": "FoundationKivaSettings", "..." : "..." }
}
```

An HBJSON with these keys opened by an *older* honeybee-energy is still a fully valid model — the faces are ordinary `Ground` faces and the extra keys are ignored.

---

## 4. Translation behavior

**Rule: a face is written with the `Foundation` BC when and only when it (a) has a `Ground` boundary condition, (b) is a Floor or Wall, and (c) has `kiva_foundation` assigned.** Everything else is untouched.

```mermaid
flowchart LR
    A[Face<br/>BC = Ground] --> B{kiva_foundation<br/>assigned?}
    B -- no --> C[IDF: Outside BC = Ground<br/><i>unchanged from today</i>]
    B -- yes --> D[IDF: Outside BC = Foundation<br/>BC Object = foundation id]
    D --> E{Floor?}
    E -- yes --> F[+ SurfaceProperty:<br/>ExposedFoundationPerimeter]
    D --> G[model level: unique Foundation:Kiva objects<br/>+ their materials + Foundation:Kiva:Settings]
```

**IDF writer** (`honeybee_energy/writer.py`):
- `face_to_idf`: new branch in the BC block (alongside the `OtherSideTemperature` branch) setting `bc_name='Foundation'` and the object reference; for Floors, the `SurfaceProperty:ExposedFoundationPerimeter` string is appended via the existing `append_txt` mechanism.
- `model_to_idf`: new section (like the HVAC/PV blocks) writing each unique `Foundation:Kiva` once + the settings object; Kiva-referenced materials/constructions join the existing dedupe-and-write materials block.

**honeybee-openstudio**: the OpenStudio SDK has complete Kiva support, and `Surface.setAdjacentFoundation(kiva)` even sets the BC string + object in one call. Plan: build an id→`OSFoundationKiva` map right after constructions in `model_to_openstudio`; in `face_to_openstudio`, call `setAdjacentFoundation` + `createSurfacePropertyExposedFoundationPerimeter` (mirroring the existing `OtherSideTemperature` branch); hard-set the construction on these faces (OS construction-set inheritance can't resolve BC="Foundation", same as the existing Adiabatic/OtherSideTemperature handling); write `FoundationKivaSettings` via `getFoundationKivaSettings()`.
- ⚠ **Release coordination**: the honeybee-openstudio PR must land before any LBT release that includes the honeybee-energy PR — otherwise the OSM path silently drops foundations while direct IDF export honors them, and users get different results from the two routes.

**Other consumers** (TRACE preprocessing, gbXML, dragonfly, loadbalance `EXTERIOR_BCS`): no changes needed — the BC is still `Ground` everywhere. This is the biggest payoff of Design C.

### Exposed perimeter auto-calculation

For each boundary segment of a foundation Floor, look for a same-room wall sharing that edge in plan (within model tolerance):

- wall is `Outdoors` or a foundation wall → segment **exposed**
- wall is `Surface` or `Adiabatic` (party wall / interior partition) → segment **not exposed**
- no wall found on that edge → **exposed** (conservative)

```
                     street side (Outdoors walls)
                 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                 ┃                               ┃
    party wall   ┋                               ┋   party wall
    (Adiabatic/  ┋        FOUNDATION SLAB        ┋   (Adiabatic/
     Surface →   ┋                               ┋    Surface →
     NOT exposed)┋                               ┋    NOT exposed)
                 ┃                               ┃
                 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                     rear yard (Outdoors walls)

                 ━━━ exposed perimeter    ┋ unexposed
```

This matters enormously for rowhouses/townhouses and multifamily, where `ExposedPerimeterFraction = 1.0` (the E+ default) badly overstates perimeter losses. The per-face `exposed_perimeter` numeric override covers cases the heuristic can't know about, and `0` is legal (core slab). If you'd rather keep v1 simpler, the autocalc could ship as a later addition with explicit values required initially — your call (open question #3).

---

## 5. Grasshopper UI — user stories & new components

Three new components in honeybee-grasshopper-energy. No changes to honeybee-grasshopper-core.

<img width="1230" height="654" alt="Image" src="https://github.com/user-attachments/assets/c9d1f28d-d935-4c16-895e-714456a0c323" />

### Component 1: `HB Kiva Foundation` — create the resource
*(modeled on `HB Other Side Temperature`; `0 :: Basic Properties`)*

```
                        ┌───────────────────────────────┐
   _name_ ─────────────►│                               │
   int_horiz_insul_ ───►│                               │
   int_horiz_depth_ ───►│                               │
   int_horiz_width_ ───►│                               │
   int_vert_insul_ ────►│                               │
   int_vert_depth_ ────►│      HB Kiva Foundation       │────► foundation
   ext_horiz_insul_ ───►│                               │
   ext_horiz_depth_ ───►│                               │
   ext_horiz_width_ ───►│                               │
   ext_vert_insul_ ────►│                               │
   ext_vert_depth_ ────►│                               │
   _wall_height_ ──────►│  (all inputs optional —       │
   _wall_depth_ ───────►│   blank = E+ defaults)        │
   footing_constr_ ────►│                               │
   footing_mat_ ───────►│                               │
   _footing_depth_ ────►│                               │
                        └───────────────────────────────┘
```

Insulation inputs accept an `EnergyMaterial` or a text identifier (resolved from the material lib — same convention as `HB Opaque Construction`). Every input optional, mirroring the E+ object where only `Name` is required. It's a wide component, but it's a 1:1 mirror of the E+ IOR, which I think beats inventing an abstraction (open question #5 if you'd prefer grouping).

### Component 2: `HB Apply Kiva Foundation` — assign to rooms/faces
*(modeled on `HB Adiabatic by Type` / `HB Apply Opaque Construction`; `0 :: Basic Properties`)*

```
                        ┌───────────────────────────────┐
   _hb_objs ───────────►│                               │
   foundation_ ────────►│   HB Apply Kiva Foundation    │────► hb_objs
   include_walls_ ─────►│                               │
   exposed_per_ ───────►│                               │
                        └───────────────────────────────┘
```

- `_hb_objs`: Rooms, Faces, or a Model (duplicated before mutation, per convention).
- `foundation_`: a `FoundationKiva` (or identifier). **If blank, a default all-defaults `FoundationKiva` is created** — the "just give me Kiva" path.
- `include_walls_`: bool (default `True`) — also assign the foundation to the room's `Ground`-BC walls so they join the 2-D model (per the E+ docs, walls and floor should share the object).
- `exposed_per_`: optional number to override autocalculate (only sensible when `_hb_objs` are individual Faces).
- Assignment logic per Room: every `Ground`-BC Floor face gets the foundation; **if a room has multiple ground floors, the component auto-duplicates the foundation per floor** (identifier-suffixed copies) to satisfy E+'s one-floor-per-Kiva-per-zone rule, and attaches each `Ground` wall to the copy belonging to the floor it shares an edge with. The core library stays strict (validation error); the component provides the convenience. (Open question #2.)
- Faces that are not `Ground` BC are skipped — the component composes with `HB Custom Ground` / room-from-solid ground detection, which have already established which faces are ground contact.

### Component 3: `HB Kiva Settings` — model-wide soil properties
*(model-level singleton, following the electric-load-center precedent; `5 :: Simulate` or `0 :: Basic Properties`)*

```
                        ┌───────────────────────────────┐
   _model ─────────────►│                               │
   _soil_conduct_ ─────►│                               │
   _soil_density_ ─────►│                               │
   _soil_spec_heat_ ───►│                               │
   _solar_abs_ ────────►│       HB Kiva Settings        │────► model
   _thermal_abs_ ──────►│                               │
   _roughness_ ────────►│  (all optional — blank =      │
   _far_field_ ────────►│   E+ defaults)                │
   _deep_ground_ ──────►│                               │
   _timestep_ ─────────►│                               │
                        └───────────────────────────────┘
```

The deep-ground inputs enforce the E+ coupling rule in the class itself (BC=`Autoselect` ⟺ depth=`autocalculate`), so users can't author an invalid pair.

### User stories

**Story 1 — "Just model my slab with Kiva" (minimum viable workflow):**

```
[HB Room from Solid] ──► [HB Apply Kiva Foundation] ──► [HB Model] ──► [HB Model to OSM]
                              (no inputs beyond _hb_objs)
```

One component insertion into an existing definition. Every ground floor gets a defaults-only `Foundation:Kiva`, exposed perimeter is auto-computed. This is the workflow 90% of users need.

**Story 2 — insulated basement (EnerPHit-style retrofit):**

```
[HB Opaque Material: "XPS 2in"] ─────────────┐
                                             ▼
                              [HB Kiva Foundation]
                               ext_vert_insul_ = XPS 2in
                               ext_vert_depth_ = 0.6
                                             │ foundation
[HB Room from Solid] ──────────────────────► ▼
 (basement below grade →      [HB Apply Kiva Foundation] ──► [HB Model] ──► ...
  Ground BC on walls+floor)    include_walls_ = True
```

Slab and basement walls keep their own (different) constructions; both reference the one foundation. Continuous under-slab insulation goes in the *floor construction*; edge/exterior insulation and footings go in the *foundation object* — the component descriptions will spell out this split since it's the classic Kiva confusion.

**Story 3 — townhouse with party walls:** identical to Story 1 or 2; the exposed-perimeter autocalc detects the `Surface`/`Adiabatic` party walls and excludes those edges (diagram in §4). No user action required — this is the case where E+'s defaults are most wrong and where this feature pays off for retrofit modeling.

**Story 4 — multifamily, many identical rooms:** one `HB Kiva Foundation` definition applied to a list of Rooms. Each room's floor references the same object identifier; E+ instantiates a separate 2-D Kiva model per zone automatically. One definition, no per-room duplication in the user's file.

---

## 6. Validation

New `check_*` methods on `ModelEnergyProperties` (wired into `check_for_extension` + `ERROR_MAP` with new `02xxxx` codes, following the SHW precedent):

| Check | Severity |
|---|---|
| Duplicate `FoundationKiva` identifiers in the model | error |
| Two floors in one room referencing the same foundation (E+ one-floor-per-Kiva-per-zone rule) | error |
| Foundation wall whose foundation isn't also on exactly one floor in the same room | error |
| `kiva_foundation` on a face that is not a `Ground`-BC Floor/Wall (incl. Outdoors faces, RoofCeilings, AirBoundaries) | error |
| Apertures/Doors on a face with `kiva_foundation` | error |
| R-only layers (`EnergyMaterialNoMass`, vegetation…) in the construction of a foundation face, its footing-wall construction, or its insulation fields | error |

## 7. Documented limitations (docs/component descriptions, no code)

- **Weather file required** — Kiva fails in design-day-only runs (its accelerated warmup reads the EPW).
- **No radiant slabs** on Kiva faces (`Construction:InternalSource` conflict) — relevant for Radiant-with-DOAS HVAC on ground floors.
- E+ warns on foundation walls with >4 vertices and expects walls to share a plan edge with their floor — worth a note, not a check.
- Zone multipliers, walkout basements (mixed wall heights), and mixed wall constructions are all handled inside E+ — no action, but worth documenting that they work.

## 8. Rollout (4 PRs, in dependency order)

1. **honeybee-energy** — `foundation.py`, face/model properties, IDF writer, validation, tests. *(this PR)*
2. **honeybee-schema** — `FoundationKiva`/`FoundationKivaSettings` classes, additive fields, sample model.
3. **honeybee-openstudio** — OSM translation (must land before the same LBT release as PR 1 — see §4).
4. **honeybee-grasshopper-energy** — the three components above (I'll export the `.ghuser`/json/icons via LB Export UserObject).

Happy to do all four; dragonfly-energy (Room2D `is_ground_contact` → foundation property) could be a follow-up once the pattern settles.

## 9. Open questions for you

1. **Design blessing** — are you comfortable with the face-property + shared-resource shape (Design C), given it's your construction-trigger idea adjusted for the wall/floor sharing rule?
2. **Multi-floor rooms** — OK with the GH component auto-duplicating foundations per floor while the core library stays strict (validation error)?
3. **Exposed perimeter autocalc** — in scope for v1, or ship explicit-values-only first and add the heuristic after?
4. **Issue #666 (C/F-factor ground BCs)** — you'd mentioned possibly doing these together. Design C extends naturally (an F/C-factor spec would be another face-level ground property rather than a new BC), but I'd propose keeping them out of PR 1 to keep the review tractable.
5. **Component granularity** — is the 16-input 1:1 `HB Kiva Foundation` component acceptable, or would you rather group the four insulation positions into a helper object/component?
6. **Naming** — `kiva_foundation` / `FoundationKiva` / `foundations` / `foundation_kiva_settings`? Went with E+-mirroring names, but happy to take your preference.

If the shape looks right to you, I'll could start on PR 1 and open it as a draft so you can take a look.

Thanks!
@ed-p-may

**Comment by chriswmackey** [comment]
Thanks, @ed-p-may .

I only made it half way through so far and I'll read the rest later. I generally agree with the proposal for how to implement it in honeybee-energy and honeybee-schema. All of the changes that I would propose involve making support for KIVA more like how we support the E+ AFN since it's very similar in principle. Essentially, the model can run with all simple objects and, if you want to use this advanced feature set of energyplus, you just pass the model through a component that assigns all of the properties of this advanced model for you, assuming the model is otherwise ready to simulate.

To this end, I would recommend the following changes to the proposal:

1. Let's not worry about exposing this on the model.to_idf() pathway right now and only focus on support with honeybee-openstudio. Model.to_idf() is really only intended for a simple load model and something as advanced as KIVA is a little out of place there. If you really want it to be a part of this method, we can do it after we get the process working with honeybee-openstudio translation. But the AFN and detailed HVAC templates are not exposed in Model.to_idf() so I think the vast majority of users will similarly be fine if KIVA is an OpenStudio-only feature.
2. For the Grasshopper components, I think there should be one component like [HBA AirflowNetwork](https://docs.ladybug.tools/hb-energy-primer/components/3_loads/airflow_newtwork) that just applies everything to the ground boundary conditions of the model for you. I know that you want to expose more but that component in your screenshot is huge and the vast majority of people who want to use KIVA are not going to touch those properties. I think it is fine if they remain as things that are customizable only on the SDK layer (like the vast majority of AFN properties).
